<?xml version="1.0"?>
<robot name="panda" xmlns:xacro="http://ros.org/wiki/xacro">
  <xacro:macro name="panda" params="">
  <xacro:property name="panda_mesh_prefix"
      value="package://panda_description/panda_model/panda_mounted_husky/panda_model_description/meshes"/>
  <!-- Paths -->
  <!-- ================== LINKS ================== -->

  <link name="panda_link0">
    <inertial>
      <origin xyz="-0.025566 -2.87883e-05 0.057332" rpy="0 0 0"/>
      <mass value="2.814271300104062"/>
      <inertia ixx="0.0075390636364814695" iyy="0.010508017427382959" izz="0.009864933638708275"
               ixy="-1.5053124574783335e-05" ixz="-0.0012144328835738973" iyz="1.0434803501032979e-05"/>
    </inertial>
    <visual><geometry><mesh filename="${panda_mesh_prefix}/visual/link0.dae"/></geometry></visual>
    <collision><geometry><mesh filename="${panda_mesh_prefix}/collision/link0.stl"/></geometry></collision>
  </link>

  <link name="panda_link1">
    <inertial>
      <origin xyz="-3.72212e-09 -0.0324958 -0.0675818" rpy="0 0 0"/>
      <mass value="2.6433170427227215"/>
      <inertia ixx="0.017405076982879724" iyy="0.015352077935781732" izz="0.005987891663073857"
               ixy="-8.687986369621078e-09" ixz="-1.6277819062920423e-08" iyz="0.004510855959102245"/>
    </inertial>
    <visual><geometry><mesh filename="${panda_mesh_prefix}/visual/link1.dae"/></geometry></visual>
    <collision><geometry><mesh filename="${panda_mesh_prefix}/collision/link1.stl"/></geometry></collision>
  </link>

  <link name="panda_link2">
    <inertial>
      <origin xyz="3.17833e-07 -0.06861 0.0322285" rpy="0 0 0"/>
      <mass value="2.6652755331661186"/>
      <inertia ixx="0.01778694005635383" iyy="0.006044133867952844" izz="0.01571095500455695"
               ixy="8.70635625879119e-08" ixz="3.752072058394008e-08" iyz="-0.004599490295490376"/>
    </inertial>
    <visual><geometry><mesh filename="${panda_mesh_prefix}/visual/link2.dae"/></geometry></visual>
    <collision><geometry><mesh filename="${panda_mesh_prefix}/collision/link2.stl"/></geometry></collision>
  </link>

  <link name="panda_link3">
    <inertial>
      <origin xyz="0.0469893 0.0316374 -0.031704" rpy="0 0 0"/>
      <mass value="2.3293714829045733"/>
      <inertia ixx="0.0075496764161139766" iyy="0.009680676080191398" izz="0.007944557422008086"
               ixy="-0.0024320464513492225" ixz="-0.003250815048103007" iyz="-0.0021316282421730694"/>
    </inertial>
    <visual><geometry><mesh filename="${panda_mesh_prefix}/visual/link3.dae"/></geometry></visual>
    <collision><geometry><mesh filename="${panda_mesh_prefix}/collision/link3.stl"/></geometry></collision>
  </link>

  <link name="panda_link4">
    <inertial>
      <origin xyz="-0.0360505 0.0337069 0.0319019" rpy="0 0 0"/>
      <mass value="2.3104763646281508"/>
      <inertia ixx="0.007764568814864828" iyy="0.008015776929980149" izz="0.009955882308071066"
               ixy="0.003371475204045935" ixz="-0.002343621113168248" iyz="0.0022917526161470624"/>
    </inertial>
    <visual><geometry><mesh filename="${panda_mesh_prefix}/visual/link4.dae"/></geometry></visual>
    <collision><geometry><mesh filename="${panda_mesh_prefix}/collision/link4.stl"/></geometry></collision>
  </link>

  <link name="panda_link5">
    <inertial>
      <origin xyz="-6.12525e-06 0.0610427 -0.104176" rpy="0 0 0"/>
      <mass value="2.6226426154892004"/>
      <inertia ixx="0.029070538014399377" iyy="0.027638529986528795" izz="0.004251111234400737"
               ixy="6.208738268725568e-07" ixz="-1.0059151925023392e-05" iyz="-0.007424306477850814"/>
    </inertial>
    <visual><geometry><mesh filename="${panda_mesh_prefix}/visual/link5.dae"/></geometry></visual>
    <collision><geometry><mesh filename="${panda_mesh_prefix}/collision/link5.stl"/></geometry></collision>
  </link>

  <link name="panda_link6">
    <inertial>
      <origin xyz="0.051051 0.00910797 0.0106343" rpy="0 0 0"/>
      <mass value="1.5025804886626795"/>
      <inertia ixx="0.002940555801469018" iyy="0.00392105201248677" izz="0.0054115603870960195"
               ixy="-0.00042391108351381507" ixz="0.000610007735338717" iyz="0.00012647545430836214"/>
    </inertial>
    <visual><geometry><mesh filename="${panda_mesh_prefix}/visual/link6.dae"/></geometry></visual>
    <collision><geometry><mesh filename="${panda_mesh_prefix}/collision/link6.stl"/></geometry></collision>
  </link>

  <link name="panda_link7">
    <inertial>
      <origin xyz="0.0109695 0.0107965 0.0650412" rpy="0 0 0"/>
      <mass value="0.5237668134788256"/>
      <inertia ixx="0.0008621467959726535" iyy="0.0008613100184801048" izz="0.0006961139396618219"
               ixy="-0.00011871071570425467" ixz="3.866602604509339e-05" iyz="-9.057841649422724e-05"/>
    </inertial>
    <visual name="panda_link7_visual">
      <geometry><mesh filename="${panda_mesh_prefix}/visual/link7.dae"/></geometry>
    </visual>
    <collision name="panda_link7_collision">
      <geometry><mesh filename="${panda_mesh_prefix}/collision/link7.stl"/></geometry>
    </collision>

    <!-- Hand shell mounted on link7 (from SDF) -->
    <collision name="panda_hand_collision">
      <origin xyz="0 0 0.107" rpy="0 0 -0.7853981633974483"/>
      <geometry><mesh filename="${panda_mesh_prefix}/collision/hand.stl" scale="1 1 1"/></geometry>
    </collision>
    <visual name="panda_hand_visual">
      <origin xyz="0 0 0.107" rpy="0 0 -0.7853981633974483"/>
      <geometry><mesh filename="${panda_mesh_prefix}/visual/hand.dae" scale="1 1 1"/></geometry>
    </visual>
  </link>

  <!-- SDF frame as massless link -->
  <link name="wrist_mount"/>
  <!-- Camera body (small mass to keep physics stable) -->
  <link name="wrist_cam_link">
    <inertial>
      <mass value="0.08"/>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <inertia ixx="0.0001" iyy="0.0001" izz="0.0001" ixy="0" ixz="0" iyz="0"/>
    </inertial>
    <visual name="wrist_cam_debug">
      <geometry><box size="0.06 0.02 0.02"/></geometry>
    </visual>
  </link>
  <!-- REP 103 optical frame -->
  <link name="wrist_cam_optical_frame"/>

  <!-- Fingers -->
  <link name="panda_leftfinger">
    <inertial>
      <origin xyz="-6.2561e-06 0.00227781 -0.0277116" rpy="0 0 0"/>
      <mass value="0.2238453537819424"/>
      <inertia ixx="6.7480416736369505e-05" iyy="6.6271892216413953e-05" izz="1.550396267804825e-05"
               ixy="-4.1027443646025684e-09" ixz="1.8693673970152991e-08" iyz="1.19675484743908e-05"/>
    </inertial>
    <visual><geometry><mesh filename="${panda_mesh_prefix}/visual/finger.dae"/></geometry></visual>
    <collision><geometry><mesh filename="${panda_mesh_prefix}/collision/finger.stl"/></geometry></collision>
  </link>

  <link name="panda_rightfinger">
    <inertial>
      <origin xyz="-6.2561e-06 0.00227781 -0.0277116" rpy="0 0 3.141592653589793"/>
      <mass value="0.2238453537819424"/>
      <inertia ixx="6.7480416736369505e-05" iyy="6.6271892216413953e-05" izz="1.550396267804825e-05"
               ixy="-4.1027443646025684e-09" ixz="1.8693673970152991e-08" iyz="1.19675484743908e-05"/>
    </inertial>
    <visual>
      <origin xyz="0 0 0" rpy="0 0 3.141592653589793"/>
      <geometry><mesh filename="${panda_mesh_prefix}/visual/finger.dae"/></geometry>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 3.141592653589793"/>
      <geometry><mesh filename="${panda_mesh_prefix}/collision/finger.stl"/></geometry>
    </collision>
  </link>

  <!-- ================== JOINTS ================== -->

  <!-- <link name="world"/>

  <joint name="world_to_panda" type="fixed">
    <parent link="world"/>
    <child  link="panda_link0"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
  </joint> -->
  <joint name="panda_joint1" type="revolute">
    <parent link="panda_link0"/><child link="panda_link1"/>
    <origin xyz="0 0 0.333" rpy="0 0 0"/>
    <axis xyz="0 0 1"/>
    <limit effort="87" velocity="2.617993877991494" lower="-2.897246558310587" upper="2.897246558310587"/>
    <dynamics damping="21.75" friction="3.48"/>
  </joint>

  <joint name="panda_joint2" type="revolute">
    <parent link="panda_link1"/><child link="panda_link2"/>
    <origin xyz="0 0 0" rpy="-1.570796326794897 0 0"/>
    <axis xyz="0 0 1"/>
    <limit effort="87" velocity="2.617993877991494" lower="-1.762782544514273" upper="1.762782544514273"/>
    <dynamics damping="21.75" friction="3.48"/>
  </joint>

  <joint name="panda_joint3" type="revolute">
    <parent link="panda_link2"/><child link="panda_link3"/>
    <origin xyz="0 -0.316 0" rpy="1.570796326794897 0 0"/>
    <axis xyz="0 0 1"/>
    <limit effort="87" velocity="2.617993877991494" lower="-2.897246558310587" upper="2.897246558310587"/>
    <dynamics damping="21.75" friction="3.48"/>
  </joint>

  <joint name="panda_joint4" type="revolute">
    <parent link="panda_link3"/><child link="panda_link4"/>
    <origin xyz="0.0825 0 0" rpy="1.570796326794897 0 0"/>
    <axis xyz="0 0 1"/>
    <limit effort="87" velocity="2.617993877991494" lower="-3.07177948351002" upper="-0.06981317007977318"/>
    <dynamics damping="21.75" friction="3.48"/>
  </joint>

  <joint name="panda_joint5" type="revolute">
    <parent link="panda_link4"/><child link="panda_link5"/>
    <origin xyz="-0.0825 0.384 0" rpy="-1.570796326794897 0 0"/>
    <axis xyz="0 0 1"/>
    <limit effort="12" velocity="3.141592653589793" lower="-2.897246558310587" upper="2.897246558310587"/>
    <dynamics damping="3.0" friction="0.48"/>
  </joint>

  <joint name="panda_joint6" type="revolute">
    <parent link="panda_link5"/><child link="panda_link6"/>
    <origin xyz="0 0 0" rpy="1.570796326794897 0 0"/>
    <axis xyz="0 0 1"/>
    <limit effort="12" velocity="3.141592653589793" lower="-0.0174532925199433" upper="3.752457891787809"/>
    <dynamics damping="3.0" friction="0.48"/>
  </joint>

  <joint name="panda_joint7" type="revolute">
    <parent link="panda_link6"/><child link="panda_link7"/>
    <origin xyz="0.088 0 0" rpy="1.570796326794897 -0.8 0.0"/>
    <axis xyz="0 0 1"/>
    <limit effort="12" velocity="3.141592653589793" lower="-2.897246558310587" upper="2.897246558310587"/>
    <dynamics damping="3.0" friction="0.48"/>
  </joint>

  <!-- fixed frames -->
  <joint name="panda_wrist_mount" type="fixed">
    <parent link="panda_link7"/><child link="wrist_mount"/>
    <origin xyz="0 0 0.107" rpy="0 0 -0.7853981633974483"/>
  </joint>

  <joint name="wrist_cam_mount" type="fixed">
    <parent link="wrist_mount"/><child link="wrist_cam_link"/>
    <!-- SDF had +5 cm forward from wrist_mount -->
    <origin xyz="0 0 0.05" rpy="0 0 0"/>
  </joint>

  <joint name="wrist_cam_optical_joint" type="fixed">
    <parent link="wrist_cam_link"/><child link="wrist_cam_optical_frame"/>
    <!-- Z forward, X right, Y down -->
    <origin xyz="0 0 0" rpy="-1.57079632679 0 -1.57079632679"/>
  </joint>

  <!-- Fingers (prismatic) -->
  <joint name="panda_finger_joint1" type="prismatic">
    <parent link="panda_link7"/><child link="panda_leftfinger"/>
    <origin xyz="0 0 0.1654" rpy="0 0 -0.7853981633974483"/>
    <axis xyz="0 1 0"/>
    <limit lower="0.0" upper="0.04" effort="20" velocity="0.2"/>
    <dynamics damping="3.0" friction="0.4"/>
  </joint>

  <joint name="panda_finger_joint2" type="prismatic">
    <parent link="panda_link7"/><child link="panda_rightfinger"/>
    <origin xyz="0 0 0.1654" rpy="0 0 -0.7853981633974483"/>
    <axis xyz="0 -1 0"/>
    <limit lower="0.0" upper="0.04" effort="20" velocity="0.2"/>
    <dynamics damping="3.0" friction="0.4"/>
  </joint>

  <!-- ================== ros2_control ================== -->
  <ros2_control name="GazeboSimSystem" type="system">
    <hardware>
      <plugin>gz_ros2_control/GazeboSimSystem</plugin>
    </hardware>

    <!-- Arm -->
    <joint name="panda_joint1">
      <command_interface name="position"/>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
    </joint>
    <joint name="panda_joint2"><command_interface name="position"/><state_interface name="position"/><state_interface name="velocity"/><state_interface name="effort"/></joint>
    <joint name="panda_joint3"><command_interface name="position"/><state_interface name="position"/><state_interface name="velocity"/><state_interface name="effort"/></joint>
    <joint name="panda_joint4"><command_interface name="position"/><state_interface name="position"/><state_interface name="velocity"/><state_interface name="effort"/></joint>
    <joint name="panda_joint5"><command_interface name="position"/><state_interface name="position"/><state_interface name="velocity"/><state_interface name="effort"/></joint>
    <joint name="panda_joint6"><command_interface name="position"/><state_interface name="position"/><state_interface name="velocity"/><state_interface name="effort"/></joint>
    <joint name="panda_joint7"><command_interface name="position"/><state_interface name="position"/><state_interface name="velocity"/><state_interface name="effort"/></joint>

    <!-- wrist fingers -->
    <joint name="panda_finger_joint1"><command_interface name="position"/><state_interface name="position"/><state_interface name="velocity"/></joint>
    <joint name="panda_finger_joint2"><command_interface name="position"/><state_interface name="position"/><state_interface name="velocity"/></joint>
  </ros2_control>

  <!-- ================== Gazebo: load ros2_control, friction, and sensor ================== -->
  <!-- <gazebo>
    <plugin filename="libgz_ros2_control-system.so" name="gz_ros2_control::GazeboSimROS2ControlPlugin">
      <parameters>/config/controllers.yaml</parameters>
    </plugin>
  </gazebo> -->

  <gazebo reference="panda_link0"><gravity>false</gravity></gazebo>

  <gazebo>
      <!-- J1 -->
    <plugin filename="libgz-sim-joint-position-controller-system.so"
            name="gz::sim::systems::JointPositionController">
      <joint_name>panda_joint1</joint_name>
      <p_gain>50</p_gain><d_gain>20</d_gain><i_gain>0</i_gain>
      <i_min>-1</i_min><i_max>1</i_max>
      <cmd_min>-87</cmd_min><cmd_max>87</cmd_max>
      <topic>/model/panda/joint/panda_joint1/cmd_pos</topic>
    </plugin>

    <!-- J2 -->
    <plugin filename="libgz-sim-joint-position-controller-system.so"
            name="gz::sim::systems::JointPositionController">
      <joint_name>panda_joint2</joint_name>
      <p_gain>10000</p_gain><d_gain>500</d_gain><i_gain>0</i_gain>
      <i_min>-1</i_min><i_max>1</i_max>
      <cmd_min>-87</cmd_min><cmd_max>87</cmd_max>
      <topic>/model/panda/joint/panda_joint2/cmd_pos</topic>
    </plugin>

    <!-- J3 -->
    <plugin filename="libgz-sim-joint-position-controller-system.so"
            name="gz::sim::systems::JointPositionController">
      <joint_name>panda_joint3</joint_name>
      <p_gain>100</p_gain><d_gain>10</d_gain><i_gain>0</i_gain>
      <i_min>-1</i_min><i_max>1</i_max>
      <cmd_min>-87</cmd_min><cmd_max>87</cmd_max>
      <topic>/model/panda/joint/panda_joint3/cmd_pos</topic>
    </plugin>

    <!-- J4 -->
    <plugin filename="libgz-sim-joint-position-controller-system.so"
            name="gz::sim::systems::JointPositionController">
      <joint_name>panda_joint4</joint_name>
      <p_gain>1000</p_gain><d_gain>50</d_gain><i_gain>0</i_gain>
      <i_min>-1</i_min><i_max>1</i_max>
      <cmd_min>-87</cmd_min><cmd_max>87</cmd_max>
      <topic>/model/panda/joint/panda_joint4/cmd_pos</topic>
    </plugin>

    <!-- J5 -->
    <plugin filename="libgz-sim-joint-position-controller-system.so"
            name="gz::sim::systems::JointPositionController">
      <joint_name>panda_joint5</joint_name>
      <p_gain>100</p_gain><d_gain>50</d_gain><i_gain>0</i_gain>
      <i_min>-1</i_min><i_max>1</i_max>
      <cmd_min>-12</cmd_min><cmd_max>12</cmd_max>
      <topic>/model/panda/joint/panda_joint5/cmd_pos</topic>
    </plugin>

    <!-- J6 -->
    <plugin filename="libgz-sim-joint-position-controller-system.so"
            name="gz::sim::systems::JointPositionController">
      <joint_name>panda_joint6</joint_name>
      <p_gain>100</p_gain><d_gain>10</d_gain><i_gain>0</i_gain>
      <i_min>-1</i_min><i_max>1</i_max>
      <cmd_min>-12</cmd_min><cmd_max>12</cmd_max>
      <topic>/model/panda/joint/panda_joint6/cmd_pos</topic>
    </plugin>

    <!-- J7 -->
    <plugin filename="libgz-sim-joint-position-controller-system.so"
            name="gz::sim::systems::JointPositionController">
      <joint_name>panda_joint7</joint_name>
      <p_gain>80</p_gain><d_gain>0.1</d_gain><i_gain>0</i_gain>
      <i_min>-1</i_min><i_max>1</i_max>
      <cmd_min>-12</cmd_min><cmd_max>12</cmd_max>
      <topic>/model/panda/joint/panda_joint7/cmd_pos</topic>
    </plugin>

    <!-- Fingers -->
    <plugin filename="libgz-sim-joint-position-controller-system.so"
            name="gz::sim::systems::JointPositionController">
      <joint_name>panda_finger_joint1</joint_name>
      <p_gain>100</p_gain><d_gain>10</d_gain><i_gain>0</i_gain>
      <i_min>-1</i_min><i_max>1</i_max>
      <cmd_min>-40</cmd_min><cmd_max>40</cmd_max>
      <topic>/model/panda/joint/panda_finger_joint1/cmd_pos</topic>
    </plugin>

    <plugin filename="libgz-sim-joint-position-controller-system.so"
            name="gz::sim::systems::JointPositionController">
      <joint_name>panda_finger_joint2</joint_name>
      <p_gain>100</p_gain><d_gain>10</d_gain><i_gain>0</i_gain>
      <i_min>-1</i_min><i_max>1</i_max>
      <cmd_min>-40</cmd_min><cmd_max>40</cmd_max>
      <topic>/model/panda/joint/panda_finger_joint2/cmd_pos</topic>
    </plugin>

    <!-- Publish all joint states from Gazebo -->
    <plugin filename="libgz-sim-joint-state-publisher-system.so"
            name="gz::sim::systems::JointStatePublisher">
      <joint_name>panda_joint1</joint_name>
      <joint_name>panda_joint2</joint_name>
      <joint_name>panda_joint3</joint_name>
      <joint_name>panda_joint4</joint_name>
      <joint_name>panda_joint5</joint_name>
      <joint_name>panda_joint6</joint_name>
      <joint_name>panda_joint7</joint_name>
      <joint_name>panda_finger_joint1</joint_name>
      <joint_name>panda_finger_joint2</joint_name>
      <!-- (topic parameter is optional; defaults are fine) -->
    </plugin>
  </gazebo>

  <!-- friction approximations from SDF -->
  <gazebo reference="panda_link0"><mu1>1.05</mu1><mu2>1.05</mu2></gazebo>
  <gazebo reference="panda_link1"><mu1>1.05</mu1><mu2>1.05</mu2></gazebo>
  <gazebo reference="panda_link2"><mu1>1.05</mu1><mu2>1.05</mu2></gazebo>
  <gazebo reference="panda_link3"><mu1>1.05</mu1><mu2>1.05</mu2></gazebo>
  <gazebo reference="panda_link4"><mu1>1.05</mu1><mu2>1.05</mu2></gazebo>
  <gazebo reference="panda_link5"><mu1>1.05</mu1><mu2>1.05</mu2></gazebo>
  <gazebo reference="panda_link6"><mu1>1.05</mu1><mu2>1.05</mu2></gazebo>
  <gazebo reference="panda_link7"><mu1>1.05</mu1><mu2>1.05</mu2></gazebo>
  <gazebo reference="panda_leftfinger"><mu1>1.13</mu1><mu2>1.13</mu2></gazebo>
  <gazebo reference="panda_rightfinger"><mu1>1.13</mu1><mu2>1.13</mu2></gazebo>

  <!-- Embed the RGB-D sensor (GZ Sim understands SDF inside <gazebo> in URDF) -->
  <gazebo reference="wrist_cam_link">
    <sensor name="wrist_rgbd" type="rgbd_camera">
      <pose>0.015 0 0 0 -1.57079632679 0</pose>
      <update_rate>30</update_rate>
      <always_on>true</always_on>
      <visualize>true</visualize>
      <topic>/panda/wrist/rgbd</topic>
      <camera>
        <horizontal_fov>1.0471975512</horizontal_fov>
        <image><width>640</width><height>480</height><format>R8G8B8</format></image>
        <clip><near>0.10</near><far>5.0</far></clip>
      </camera>
    </sensor>
  </gazebo>

  </xacro:macro>
</robot>
